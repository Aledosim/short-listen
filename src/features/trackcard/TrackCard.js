import React from 'react'
import styled from 'styled-components'

import playIcon from './images/play/play.png'
import infoIcon from './images/info/info.png'
import favIcon from './images/favorite/favorite.png'

export function songTime(totalTime) {
  const min = Math.floor(totalTime / 60)
  const sec = totalTime % 60

  return `${min}:${String(sec).padStart(2, "0")}`
}

// color definitions
const background = "#57287B"
const text = "#E5E5E5"
const buttonColor = "#8D46C3"
const margin = "#FFFFA7"
const border = "#ADADAD"

const Container = styled.div`
  width: 35rem;
  height: 10rem;
  display: flex;
  justify-content: space-between;
  align-items: stretch;

  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  padding-left: 1rem;
  padding-right: 1rem;

  margin-top: 1rem;
  margin-bottom: 1rem;
  margin-left: 1rem;
  margin-right: 1rem;

  border-radius: 18px;

  box-shadow: 1px 1px 5px 1px ${margin};

  background-color: ${background};
`

const Img = styled.img`
  border-radius: 5px;
  border-width: 0.5px;
  border-color: ${border};
  border-style: solid;
`
const Info = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  padding-left: 1rem;
  padding-right: 1rem;

  color: ${text};
  line-height: 2rem;
`

const ButtonContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
`

const Button = styled.button`
  height: 3.4rem;
  width: 4.7rem;
  padding-right: 10px;

  background-color: ${buttonColor};
  background-repeat: no-repeat;
  background-position: center;

  border-style: solid;
  border-radius: 5px;
  border-width: 1px;
  border-color: ${border};
`

const PlayPauseButton = styled(Button)`
  background-image: url(${props => props.icon});
`

const InfoButton = styled(Button)`
  background-image: url(${infoIcon});
`

const FavButton = styled(Button)`
  background-image: url(${favIcon});
`

export default function TrackCard(props){

  return(
    <Container role="listitem">
      <Img alt="Album cover" src={props.cover} />
      <Info>
        <div>{props.title}</div>
        <div data-cy="artistName">{props.singer}</div>
        <div>{songTime(props.time)}</div>
      </Info>
      <ButtonContainer>
        <PlayPauseButton name='play/pause' icon={playIcon}/>
        <InfoButton name='more info' />
        <FavButton name='favorite' />
      </ButtonContainer>
    </Container>
  );
};

