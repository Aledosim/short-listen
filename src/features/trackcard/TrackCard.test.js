import React from 'react'
import { render, screen } from '@testing-library/react'

import TrackCard, {songTime} from './TrackCard'

function testRender() {
    return render(
      <TrackCard
        cover="https://api.deezer.com/album/234349272/image"
        title="Meu Pedaço de Pecado"
        singer="João Gomes"
        time={157}
        preview="https://cdns-preview-7.dzcdn.net/stream/c-70255a40b7c438c3239e94ba0c909128-3.mp3"
        link="https://www.deezer.com/track/1391349252"
        id={1391349252}
      />
    )
}

describe('songTime tests', () => {

  test.each`
  time   | expected
  ${157} | ${"2:37"}
  ${60}  | ${"1:00"}
  ${1}   | ${"0:01"}
  ${600} | ${"10:00"}
  ${757} | ${"12:37"}
  ${6000}| ${"100:00"}
  `('returns $expected when $time is passed', ({time, expected}) => {
    expect(songTime(time)).toBe(expected)
  })
})

describe('<TrackCard /> tests', () => {

  it('renders without crashing', () => {
    const tree = testRender()

    expect(tree).toMatchSnapshot()
  })

  it('should show album cover, song title, singer and duration', () => {

    testRender()

    expect(screen.getByAltText(/cover/i).src).toBe("https://api.deezer.com/album/234349272/image")
    expect(screen.getByText(/Meu Pedaço de Pecado/))
    expect(screen.getByText(/João Gomes/))
    expect(screen.getByText(/2:37/))
  })

  it('should contain preview, info link and favorite buttons', () => {

    testRender()

    const buttons = screen.getAllByRole("button")
    expect(buttons.map((button) => button.getAttribute("name")))
      .toEqual(expect.arrayContaining(
        [
          "play/pause",
          "more info",
          "favorite"
        ]
      ))
    expect(buttons.length).toEqual(3)
  })

})

