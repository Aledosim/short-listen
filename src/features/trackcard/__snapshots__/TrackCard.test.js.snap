// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackCard /> tests renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  width: 35rem;
  height: 10rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  padding-left: 1rem;
  padding-right: 1rem;
  margin-top: 1rem;
  margin-bottom: 1rem;
  margin-left: 1rem;
  margin-right: 1rem;
  border-radius: 18px;
  box-shadow: 1px 1px 5px 1px #FFFFA7;
  background-color: #14213D;
}

.c1 {
  border-radius: 5px;
  border-width: 0.5px;
  border-color: #FCA311;
  border-style: solid;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-left: 1rem;
  padding-right: 1rem;
  color: #E5E5E5;
  line-height: 2rem;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
}

.c4 {
  height: 3.4rem;
  width: 4.7rem;
  padding-right: 10px;
  background-color: #2D4B8B;
  background-repeat: no-repeat;
  background-position: center;
  border-style: solid;
  border-radius: 5px;
  border-width: 1px;
  border-color: #FCA311;
}

.c5 {
  background-image: url(play.png);
}

<body>
    <div>
      <div
        class="c0"
        role="listitem"
      >
        <img
          alt="Album cover"
          class="c1"
          src="https://api.deezer.com/album/234349272/image"
        />
        <div
          class="c2"
        >
          <div>
            Meu Pedaço de Pecado
          </div>
          <div
            data-cy="artistName"
          >
            João Gomes
          </div>
          <div>
            2:37
          </div>
        </div>
        <div
          class="c3"
        >
          <button
            class="c4 c5"
            name="play/pause"
          />
          <button
            class="c4"
            name="more info"
          />
          <button
            class="c4"
            name="favorite"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdnxRM jHgyFr"
      role="listitem"
    >
      <img
        alt="Album cover"
        class="sc-gtsrHT bfbqIY"
        src="https://api.deezer.com/album/234349272/image"
      />
      <div
        class="sc-dlnjwi kuHGNG"
      >
        <div>
          Meu Pedaço de Pecado
        </div>
        <div
          data-cy="artistName"
        >
          João Gomes
        </div>
        <div>
          2:37
        </div>
      </div>
      <div
        class="sc-hKFxyN MeTXB"
      >
        <button
          class="sc-eCApnc sc-jSFjdj ivfXAQ crPzXw"
          name="play/pause"
        />
        <button
          class="sc-eCApnc ivfXAQ"
          name="more info"
        />
        <button
          class="sc-eCApnc ivfXAQ"
          name="favorite"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
